### ISSUE TITLE
Modal Accessibility

### ISSUE THREAD

**Issue by citelao** [zurb/foundation#6579]
Hi!

Your Reveal modal framework is close to being accessible, but it's missing a couple things.

1. **Focus the dialog** itself when it opens. If you `.focus()` the `role="dialog"` element, VoiceOver will do nice things, like read the title and description.
2. **Prevent tabbing outside** the modal. This requires [some hackery](https://github.com/gdkraus/accessible-modal-dialog/blob/master/modal-window.js#L38), but if you enumerate the focusable elements inside the modal you can loop around at the edges.
3. **Use a `<button>` to close**. Don't use an anchor in the examples.

That should be enough to make it accessible!

**Comment by Owlbertz** [comment]
Hi,

I was able to fix some of the points you addressed quickly.
First, I added the focus.
Second, I made the close-anchor listen to keydown events and also adjusted the examples, so they now should have a tabindex attribute to make them focusable. Of course adding href="#" would work, too, but I find tabindex better because it represents the action behavior a bit better in my opinion (abusing the anchor as a button). I also added role="button" and aria-haspopup="true". I did not fully substitute the anchor with an actual button, because this would also involve changes within the SASS which I am currently not able to perform. But I think my proposed solution will do the magic as well.
I opened a PR (#6599) for those changes.

Regarding your second point: I will have a look at this. Seems useful to me.

Regards
Marius

**Comment by gerardkcohen** [comment]
It would be so much cleaner to just a button instead of making an anchor act like a button. Focus, keyboard, semantics already handle by <button>

**Comment by citelao** [comment]
Thanks for your help!

**Comment by Owlbertz** [comment]
@gerardkcohen and @citelao : Thanks for your feedback. I will take care of this today evening or tomorrow.
(I'd suppose to use this issue for any further general communication since I find the comments within the PR a bit hard to follow...)

**Comment by citelao** [comment]
OK, I'll copy-paste my comment:


> :+1:
> We should probably include an accessibility writeup in the doc and a focus-eater extension to the js.
>
> As @gerardkcohen said, changing the <a> to a <button> is important. I am able to take a look at the SASS if need be. Were there observable side-effects on the change?

**Comment by Owlbertz** [comment]
I included the feedback and substituted the closing anchors in the examples with buttons. Turned out that there was not that much to adjust in the SASS.
I will have a look at the focus-looping suggested above soon.

**Comment by Owlbertz** [comment]
Found a solution to keep the focus within the modal

**Comment by citelao** [comment]
I don't think I can contribute to your pull request, otherwise I'd make the changes myself, but

- `aria-popup` should definitely be removed from the modal activation links. [It's only for context menus](http://www.w3.org/TR/wai-aria/states_and_properties#aria-haspopup)
- `visibleFocusableElements` should have `*[tabindex]:not([tabindex=0])` instead of just `*[tabindex]`

I can write up some accessibility notes to put in the documentation, too.

**Comment by Owlbertz** [comment]
I think `visibleFocusableElements` should include those elements with a `tabindex` of 0, since they are part of the tabbing-order. I think you mean those with a `tabindex` < 0?

**Comment by citelao** [comment]
Yeah, `tabindex="-1"`, whoops

**Comment by Owlbertz** [comment]
Btw: I updated the PR once again. It became quiet here so I think you didn't notice. Please have another look.

**Comment by citelao** [comment]
lgtm! Thanks!

**Comment by rafibomb** [comment]
@Owlbertz I really think this is a valuable discussion. You mentioned a PR. Unless I missed it, I don't see you PR merged. Did you close it? Would love to see it!

**Comment by rafibomb** [comment]
Ok found the PR - we're testing it out - thanks for submitting it!

**Comment by rafibomb** [comment]
Merged the PR!
